Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp

```diff
@@ -0,0 +1,372 @@
+        cc[2] = name[2];
+        cc[3] = name[3];
+    }
+
+    bool operator==(FourCC const&) const = default;
+    bool operator!=(FourCC const&) const = default;
+
+    char cc[4];
+};
+
+struct IFF_FILE_HEADER {
+    FourCC form;
+    BigEndian<u32> file_size;
+    FourCC format;
+};
+
+static_assert(AssertSize<IFF_FILE_HEADER, 12>());
+
+struct Chunk {
+    FourCC type;
+    ReadonlyBytes data;
+};
+
+enum CompressionType : u8 {
+    Cmp_None = 0,
+    Cmp_ByteRun = 1
+};
+
+enum MaskType : u8 {
+    Mask_None = 0,
+    Mask_HasMask = 1,
+    Mask_HasTransparentColor = 2,
+    Mask_HasLasso = 3
+};
+
+struct ChunkHeader {
+    FourCC chunk_type;
+    BigEndian<u32> chunk_size;
+};
+
+struct BitmapHeader {
+    BigEndian<u16> width;
+    BigEndian<u16> height;
+    BigEndian<i16> x;
+    BigEndian<i16> y;
+    u8 planes;
+    MaskType mask;
+    CompressionType compression;
+    u8 pad;
+    BigEndian<u16> transparent_color;
+    u8 x_aspect;
+    u8 y_aspect;
+    BigEndian<u16> page_width;
+    BigEndian<u16> page_height;
+};
+
+static_assert(sizeof(BitmapHeader) == 20);
+
+struct ILBMLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
+        BitmapHeaderDecoded,
+        FirstChunkDecoded,
+        ChunksDecoded,
+        FrameDecoded,
+        BitmapDecoded
+    };
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+
+    // points to current chunk
+    ReadonlyBytes chunks_cursor;
+
+    // max number of bytes per plane row
+    u16 pitch;
+
+    // Current file color_table in BGRA8888 format
+    Vector<u32> color_table;
+
+    ByteBuffer pixel_data;
+
+    RefPtr<Gfx::Bitmap> bitmap;
+
+    // BMHD IFF-ILBM header
+    BitmapHeader bm_header;
+
+    Optional<IntSize> size;
+};
+
+static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
+{
+    if (context.state >= ILBMLoadingContext::HeaderDecoded)
+        return {};
+
+    if (context.data.size() < sizeof(IFF_FILE_HEADER))
+        return Error::from_string_literal("Missing IFF header");
+
+    auto& header = *bit_cast<IFF_FILE_HEADER const*>(context.data.data());
+    if (header.form != FourCC("FORM") || header.format != FourCC("ILBM"))
+        return Error::from_string_literal("Invalid IFF-ILBM header");
+
+    return {};
+}
+
+static ErrorOr<Vector<u32>> decode_cmap_chunk(Chunk cmap_chunk)
+{
+    auto data = cmap_chunk.data;
+    const u8 size = (u8)(data.size() / 3);
+    Vector<u32> color_table;
+
+    for (int i = 0; i < size; ++i) {
+        u32 color = (data[i * 3] << 16) | (data[(i * 3) + 1] << 8) | data[(i * 3) + 2];
+        color |= 0xff000000;
+        color_table.append(color);
+    }
+
+    return color_table;
+}
+
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context)
+{
+    auto width = context.bm_header.width;
+    auto height = context.bm_header.height;
+    auto color_table = context.color_table;
+    auto chunky = context.pixel_data;
```

Comment:
In a general manner, I'm not a fan of these aliases. I think that they hide information (in fact, where does the information come from) just to save a few characters. But that's more a question of style, so do what you want.
However, you are copying the whole bitmap here, and that's something I can argue on being bad :)

Reply:
It's not only to save characters but also to save some indirections: I guess that's a bad habit from the JS world. Not sure that's needed in Cpp?

Reply:
1. Then you're doing premature optimization, don't try to optimize stuff without needing it and measuring it. Having a readable code is necessary, optimizing it comes after.
2. C++ compilers are very good at optimizing stuff. By doing stuff like this, you have more chance to mess with their optimizations than to outsmart them.
3. Using `auto&` would create a local reference (the size of a pointer) and that's the closest to what you want. Using `auto` and `=` calls the copy constructor which makes a __deep__ copy, copying a whole image, pixel by pixel, to save an indirection probably won't make you save some time.